c/uca/sifter/ucd.cpp

```diff
@@ -0,0 +1,180 @@
+// © 2024 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+#include "ucd.hpp"
+
+namespace unicode {
+
+CharacterDatabase::CharacterDatabase(const std::string_view versionDirectory) {
+    const auto ucdDirectory = std::filesystem::current_path()
+                                      .parent_path()
+                                      .parent_path()
+                                      .parent_path() /
+                              "unicodetools" / "data" / "ucd" /
+                              versionDirectory;
+    {
+        std::ifstream unicodeData(ucdDirectory / "UnicodeData.txt");
+        CHECK(unicodeData.good())
+                << "Run this tool from the c/uca/sifter directory";
+        std::cout << "Reading UnicodeData.txt...\n";
+        for (std::string line; std::getline(unicodeData, line);) {
+            // See https://www.unicode.org/reports/tr44/#UnicodeData.txt.
+            const auto [cp,
+                        name,
+                        gc,
+                        ccc,
+                        bc,
+                        dtDm,
+                        nvDecimal,
+                        nvDigit,
+                        nvNumeric,
+                        bm,
+                        unicode1Name,
+                        isoComment,
+                        suc,
+                        slc,
+                        stc] = fields<15>(line);
+            const char32_t codePoint = parseHexCodePoint(cp);
+            auto range = CodePointRange::inclusive(codePoint, codePoint);
+            if (name.ends_with(", First>")) {
+                std::string nextLine;
+                std::getline(unicodeData, nextLine);
+                const auto [lastCp, lastName] = fields<2>(nextLine);
+                CHECK(lastName.ends_with(", Last>")) << line << "\n"
+                                                     << nextLine;
+                CHECK(lastName.starts_with(name.substr(0, name.size() - 8)))
+                        << line << "\n"
+                        << nextLine;
+                CHECK(line.substr(cp.size() + 1 + name.size()) ==
+                      nextLine.substr(lastCp.size() + 1 + lastName.size()))
+                        << line << "\n"
+                        << nextLine;
+                range = CodePointRange::inclusive(codePoint,
+                                                  parseHexCodePoint(lastCp));
+            }
+            coarseGeneralCategory_[gc.front()].addAll(range);
+            generalCategory_[std::string(gc)].addAll(range);
+            for (char32_t c : range) {
+                canonicalCombiningClass_.emplace(c,
+                                                 std::stoi(std::string(ccc)));
```

Comment:
optional: this should fit onto one line; allow 100 chars per line?

Reply:
OBE.